### PR TITLE
Support H.265/HEVC, VP8, VP9

### PR DESCRIPTION
**Pull request by cyberj0g**
Issue link: https://github.com/livepeer/go-livepeer/issues/2123

Details:
* Added support for H.265 codecs: libx265 (CPU, encoding and decoding), NVENC (GPU, encoding and decoding)
* Added support for VP8/VP9 codecs: libvpx (CPU, encoding and decoding), NVENC (GPU, decoding only, where supported)
* Added transcoding test for all combinations of codecs and acceleration

Testing:
* install dependencies using [install_ffmpeg.sh](https://github.com/livepeer/go-livepeer/blob/ip/hevc/install_ffmpeg.sh) (WIP, please report errors)
* run tests:
  ```
  go test -v --tags=nvidia -timeout 30m -run TestNvidia_Transcoding ./ffmpeg/
  go test -v --tags=nvidia -timeout 30m -run TestSW_Transcoding ./ffmpeg/
  ```
They will output lines like below, and fail, if anything is not supported (on CI, only CPU test will run for now). SW test should work anywhere, Nvidia test should pass on GPUs starting from desktop GTX 1070 - it can't decode VP9 on my GTX 1070m, despite support matrix stating it should.
```
    ffmpeg_test.go:754: Transcode VP8 (Accel: 0) -> H.265/HEVC (Accel: 0) Prepare: true Transcode: true Quality: true
    ffmpeg_test.go:754: Transcode VP8 (Accel: 0) -> VP8 (Accel: 0) Prepare: true Transcode: true Quality: true
    ffmpeg_test.go:754: Transcode VP8 (Accel: 0) -> VP9 (Accel: 0) Prepare: true Transcode: true Quality: true
```

### CODE REVIEW

**Comment by cyberj0g** [comment]
Finished testing new 'unified' [install_ffmpeg.sh](https://github.com/livepeer/go-livepeer/blob/ip/hevc/install_ffmpeg.sh) from go-livepeer and added it to circle-ci job. We shouldn't need a separate one for lpms now. For dev/test install, add `debug-video` to BUILD_TAGS e.g.:
```
BUILD_TAGS="experimental" ./install_ffmpeg.sh /projects/livepeer/src/
```
I checked new codecs tests with this `install_ffmpeg.sh`, but not sure if all other tests are fine. @JamesWanglf is welcome to add his changes and create PR in go-livepeer.

**Comment by leszko** [comment]
> @leszko What is the reason for including `matroska,opus,webm,webm_chunk,webm_dash_manifest` into production build?

hmm? What do you mean?

**Comment by darkdarkdragon** [comment]
> > @leszko What is the reason for including `matroska,opus,webm,webm_chunk,webm_dash_manifest` into production build?
> 
> hmm? What do you mean?

In install_ffmpeg.sh you enabled those muxers by default, which means that they will go into production `go-livepeer` build.

**Comment by darkdarkdragon** [comment]
> > > > @leszko What is the reason for including `matroska,opus,webm,webm_chunk,webm_dash_manifest` into production build?
> > > 
> > > 
> > > hmm? What do you mean?
> > 
> > 
> > In install_ffmpeg.sh you enabled those muxers by default, which means that they will go into production `go-livepeer` build.
> 
> Did I? Where is this change?

Sorry, it was not you 😄

**Comment by cyberj0g** [comment]
@darkdarkdragon I added these. It makes sense to support MKV and WebM containers for VP8/9, and AAC is not supported for WebM - that's why Opus is added. DASH is a great alternative to HLS (especially for WebM), we might want to support it in the near future too.

**Comment by leszko** [comment]
> > > > > @leszko What is the reason for including `matroska,opus,webm,webm_chunk,webm_dash_manifest` into production build?
> > > > 
> > > > 
> > > > hmm? What do you mean?
> > > 
> > > 
> > > In install_ffmpeg.sh you enabled those muxers by default, which means that they will go into production `go-livepeer` build.
> > 
> > 
> > Did I? Where is this change?
> 
> Sorry, it was not you 😄

🙃

**Comment by darkdarkdragon** [comment]
> @darkdarkdragon I added these. It makes sense to support MKV and WebM containers for VP8/9, and AAC is not supported for WebM - that's why Opus is added. DASH is a great alternative to HLS (especially for WebM), we might want to support it in the near future too.

Any addition to the production build increases attack surface - so we shouldn't add anything that don't officially support. 
So I think we should remove `webm_dash_manifest` and `matroska`, and create new issue in `go-livepeer` repo for adding support for `.webm` segments.

Also, shouldn't `--enable-libopus` be added for `opus` support?

**Comment by cyberj0g** [comment]
> Any addition to the production build increases attack surface - so we shouldn't add anything that don't officially support. 

That's a good practice, but aren't we adding VP8/9 support at least for decoding with this PR? If so, common containers used with these are MKV and WEBM, and lpms now supports them. Or, since VP8/9 is an exotic choice for HLS+MPEG-TS, we should not enable these codecs at all until other streaming protocols are not implemented on go-livepeer side?

**Comment by darkdarkdragon** [comment]
> 

I think `webm` would be enough? But it's better to create issue in go-livepeer to discuss what containers should we support.

> Or, since VP8/9 is an exotic choice for HLS+MPEG-TS, we should not enable these codecs at all until other streaming protocols are not implemented on go-livepeer side?

We're really not using HLS output of go-livepeer node nowdays - it is being handled by Mist (I think in the near future, when we will integrate Mist and go-livepeer better we'll remove RTMP and HLS support from go-livepeer node)
So it is enough for go-livepeer node to being able to take .webm file and respond back with .mp4 or .ts (or .webm).

**Comment by cyberj0g** [comment]
Created a [new issue](https://github.com/livepeer/lpms/pull/279) for additional container formats, and moved these flags to debug build for now.

**Review by oscar-davids — COMMENTED** [review]
Overall LGTM!

As a separate matter, one thing I'm thinking of is to statically link x265 and libvpx like x264. Because if we link it dynamically, we need to set LD_LIBRARY_PATH or add related .so files to the distribution.  WDYT?

and I wonder how this [part ](https://github.com/livepeer/lpms/blob/0999e7e2f16c2065a6f6d2eb5b94411bfe911656/ffmpeg/ffmpeg.go#L398)will work on the VP8/VP9/H265.

**Comment by cyberj0g** [comment]
> one thing I'm thinking of is to statically link x265 and libvpx like x264

Spent quite a bit of time and got Ffmpeg to [build with new codecs as static libraries](https://github.com/livepeer/go-livepeer/blob/ip/new-codecs-static/install_ffmpeg.sh). However, libx265 is a C++ lib, and I couldn't get CGO to build lpms properly when it statically linked, and don't know if it's a good idea to switch to g++ in lpms. At this point, I believe it's not worth the effort to get it to build statically with libx265. If someone is willing to continue looking into that, my efforts stopped there (set proper PKG_CONFIG_PATH and use linked `install_ffmpeg.sh`):
```
CGO_CFLAGS="-Wno-error" CGO_CPPFLAGS="-Wno-error" CC=g++ go build -a cmd/transcoding/transcoding.go
# runtime/cgo
In file included from gcc_context.c:8:
libcgo.h:97:15: error: unnecessary parentheses in declaration of ‘_cgo_get_context_function’ [-Werror=parentheses]
   97 | extern void (*(_cgo_get_context_function(void)))(struct context_arg*);
      |               ^
cc1plus: all warnings being treated as errors
```
Without `CC=g++`, I'm getting:
```
/usr/bin/ld: /projects/livepeer/src/compiled/lib/libx265.a(encoder.cpp.o): undefined reference to symbol '_ZTVN10__cxxabiv121__vmi_class_type_infoE@@CXXABI_1.3'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu//libstdc++.so.6: error adding symbols: DSO missing from command line
```

**Comment by cyberj0g** [comment]
> and I wonder how this part will work on the VP8/VP9/H265.

It doesn't, none of these parameters work for HEVC or VP8/9 to my knowledge.

**Comment by iameli** [comment]
@cyberj0g Does the node boot without libx265, only loading it dynamically when needed? I don't think we probably want to add an external libx265 dependency if not.

**Comment by cyberj0g** [comment]
> @cyberj0g Does the node boot without libx265, only loading it dynamically when needed? I don't think we probably want to add an external libx265 dependency if not.

I think we agreed to not include software versions of new codecs into production build for now, so it won't be an issue. In debug mode, `--enable-shared` is added, so all dependencies are linked dynamically. As an experiment, I quickly tried building without `--enable-shared` with all dependencies static, except `libx265.so` - it didn't work, probably something is missing from Ffmpeg's `configure.sh` to support mixing static and shared dependencies.

File: ffmpeg/ffmpeg_test.go
Comment (yondonfu) [inline ffmpeg/ffmpeg_test.go:31]: Do we need to return a bool here if the cmd string executed via `exec.Command()` triggers an error if it ends with an exit code of 1?

See [this test example](https://github.com/livepeer/lpms/blob/5c05cbbeb53bf7cfbc413c5b33ddadf3d3699719/ffmpeg/ffmpeg_test.go#L107) that I believes will cause the test to fail if the bash command using ffprobe has an exit code of 1.

File: ffmpeg/ffmpeg_test.go
Comment (yondonfu) [inline ffmpeg/ffmpeg_test.go:696]: [t.Error](https://pkg.go.dev/testing#T.Error) is equivalent to `t.Log` followed by `t.Fail` so the `t.Fall` call should not be necessary.

File: ffmpeg/ffmpeg_test.go
Comment (yondonfu) [inline ffmpeg/ffmpeg_test.go:613]: Is this ever not true?

File: ffmpeg/ffmpeg_test.go
Comment (yondonfu) [inline ffmpeg/ffmpeg_test.go:652]: This probably doesn't need to be exported i.e. could be `codecsComboTest`.

File: ffmpeg/ffmpeg_test.go
Comment (yondonfu) [inline ffmpeg/ffmpeg_test.go:747]: Shouldn't the cmd trigger a `t.Error()` if it returns a non-zero exit code?

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: ffmpeg/decoder.c
Comment (yondonfu) [inline ffmpeg/decoder.c:245]: WDYT about adding a unit test for this i.e. passing in a non-supported codec for the decoder?

File: ffmpeg/ffmpeg_test.go
Comment (cyberj0g) [reply ffmpeg/ffmpeg_test.go:747]: I think the func produced by `setupTest` never checks exit code, it will trigger error if [something was written to stderr](https://github.com/livepeer/lpms/blob/5c05cbbeb53bf7cfbc413c5b33ddadf3d3699719/ffmpeg/ffmpeg_test.go#L33), which won't be the case here.

Comment (cyberj0g) [reply ffmpeg/ffmpeg_test.go]: Fixed.

File: ffmpeg/ffmpeg_test.go
Comment (cyberj0g) [reply ffmpeg/ffmpeg_test.go]: Removed.

**Review by cyberj0g — COMMENTED** [review]
Thanks for review @yondonfu, I addressed all comments.

File: ffmpeg/decoder.c
Comment (yondonfu) [reply ffmpeg/decoder.c:245]: Ah I see. Noticed [this comment](https://github.com/livepeer/lpms/blob/8f053cf92946f36cb2873a896b0db06713125142/ffmpeg/nvidia_test.go#L17) that should be updated then.

File: ffmpeg/ffmpeg_test.go
Comment (yondonfu) [reply ffmpeg/ffmpeg_test.go:31]: Ah right so `t.Error()` will log the error, but if the caller of the command wants to know if the command failed it currently doesn't have a way to know that and returning a bool solves that. Got it.

**Review by yondonfu — APPROVED** [review]
LGTM! I'd like to get an approval from another set of eyes that has familiarity with this codebase as well. cc @oscar-davids since you already left an initial review.

I left a minor comment about updating a comment - feel free to address prior to merging. Also, I recommend rebasing off master to pick up a fix that landed recently that should address the failing test in CI and to squash the commits to clean up the commit history before merging.

File: ffmpeg/encoder.c
Comment (yondonfu) [reply ffmpeg/encoder.c:235]: Gotcha. Seems reasonable, but I'm curious if this has resulted in any issues for anyone.

cc @iameli - have we seen issues with bitrate before?

**Review by darkdarkdragon — APPROVED** [review]
LGTM

**Review by oscar-davids — APPROVED** [review]
LGTM!
please rebase on master before merging.

**Pull request merged**
